suc/pkg/service/service.go

```diff
@@ -0,0 +1,137 @@
+package service
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/rancher/wins/pkg/defaults"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/windows/svc"
+	"golang.org/x/sys/windows/svc/mgr"
+)
+
+func RefreshWinsService() error {
+	logrus.Infof("Restarting the %s service", defaults.WindowsServiceDisplayName)
+	winSrv, err := openService(defaults.WindowsServiceName)
+	if err != nil {
+		logrus.Errorf("Cannot restart %s as the service failed to open: %v", defaults.WindowsServiceName, err)
+		return fmt.Errorf("failed to refresh the %s service: %w", winSrv.Name, err)
+	}
+	defer winSrv.Close()
+
+	state, err := getServiceState(winSrv)
+	if err != nil {
+		logrus.Errorf("Could not determine the state of the %s service: %v", defaults.WindowsServiceName, err)
+		return fmt.Errorf("could not determine the state of the %s service: %v", defaults.WindowsServiceName, err)
+	}
+
+	switch state {
+	case svc.Running:
+		err = restartService(winSrv)
+		if err != nil {
+			return fmt.Errorf("failed to refresh the %s service: %w", winSrv.Name, err)
+		}
+	case svc.Stopped:
+		err = winSrv.Start()
+		if err != nil {
+			return fmt.Errorf("failed to start %s: %w", winSrv.Name, err)
+		}
+	default:
+		logrus.Warnf("Unexpected service state %s (%d), will not start or restart service: ", serviceStateToString(state), state)
+	}
+
+	return nil
+}
+
+func getServiceState(s *mgr.Service) (svc.State, error) {
+	q, err := s.Query()
+	if err != nil {
+		return 0, fmt.Errorf("could not query service %s: %v", s.Name, err)
+	}
+
+	return q.State, nil
+}
+
+func restartService(s *mgr.Service) error {
+	currentState, err := getServiceState(s)
+	if err != nil {
+		return fmt.Errorf("failed to get state of service '%s': %w", s.Name, err)
+	}
+
+	if currentState != svc.Running {
+		// can't restart a service that isn't running
+		logrus.Warnf("Cannot restart the %s service as it is not yet running", s.Name)
+		return fmt.Errorf("cannot restart the %s service as it is not yet running", s.Name)
+	}
```

Comment:
As the goal of the function is to ensure the service is restarted, is it better to continue to start the service, instead of erroring out, in this case?

Reply:
My main goal was to mimic the `Restart-Service` powershell commandlet, which behaves in a similar way. This was also a requirement for sending the restart signal to the SCM. However, iirc in testing I saw that sending the restart signal via the windows package did not function in the same way as it does on Powershell, so I moved to explicit stop and start calls. 

For the purposes of handling the agent tls mode feature I agree, we can remove this conditional. Though there could be value in detecting when the service is not in the state we expect ('running') I don't think we will hit a case like that in the near future